bee-tangle/src/tangle.rs

```diff
@@ -0,0 +1,535 @@
+    async fn insert_approver(&self, message_id: MessageId, approver: MessageId) -> Result<(), Self::Error>;
+    /// Update the approvers list for a given message.
+    async fn update_approvers(&self, message_id: MessageId, approvers: &[MessageId]) -> Result<(), Self::Error>;
+}
+
+/// Phoney default hooks that do nothing.
+pub struct NullHooks<T>(PhantomData<T>);
+
+impl<T> Default for NullHooks<T> {
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+#[async_trait]
+impl<T: Send + Sync> Hooks<T> for NullHooks<T> {
+    type Error = ();
+
+    async fn get(&self, _message_id: &MessageId) -> Result<Option<(Message, T)>, Self::Error> {
+        Ok(None)
+    }
+
+    async fn insert(&self, _message_id: MessageId, _tx: Message, _metadata: T) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    async fn fetch_approvers(&self, _message_id: &MessageId) -> Result<Option<Vec<MessageId>>, Self::Error> {
+        Ok(None)
+    }
+
+    async fn insert_approver(&self, _message_id: MessageId, _approver: MessageId) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    async fn update_approvers(&self, _message_id: MessageId, _approvers: &[MessageId]) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+/// A foundational, thread-safe graph datastructure to represent the IOTA Tangle.
+pub struct Tangle<T, H = NullHooks<T>>
+where
+    T: Clone,
+{
+    // Global Tangle Lock. Remove this as and when it is deemed correct to do so.
+    vertices: TRwLock<HashMap<MessageId, Vertex<T>>>,
+
+    pub(crate) cache_queue: Mutex<LruCache<MessageId, (), DefaultHashBuilder>>,
+    max_len: AtomicUsize,
+
+    pub(crate) hooks: H,
+}
+
+impl<T, H: Hooks<T>> Default for Tangle<T, H>
+where
+    T: Clone,
+    H: Default,
+{
+    fn default() -> Self {
+        Self::new(H::default())
+    }
+}
+
+impl<T, H: Hooks<T>> Tangle<T, H>
+where
+    T: Clone,
+{
+    /// Creates a new Tangle.
+    pub fn new(hooks: H) -> Self {
+        Self {
+            vertices: TRwLock::new(HashMap::new()),
+
+            cache_queue: Mutex::new(LruCache::unbounded_with_hasher(DefaultHashBuilder::default())),
+            max_len: AtomicUsize::new(DEFAULT_CACHE_LEN),
+
+            hooks,
+        }
+    }
+
+    /// Create a new tangle with the given capacity.
+    pub fn with_capacity(self, cap: usize) -> Self {
+        Self {
+            cache_queue: Mutex::new(LruCache::with_hasher(cap + 1, DefaultHashBuilder::default())),
+            ..self
+        }
+    }
+
+    /// Change the maximum number of entries to store in the cache.
+    pub fn resize(&self, len: usize) {
+        self.max_len.store(len, Ordering::Relaxed);
+    }
+
+    /// Return a reference to the storage hooks used by this tangle.
+    pub fn hooks(&self) -> &H {
+        &self.hooks
+    }
+
+    async fn insert_inner(
+        &self,
+        message_id: MessageId,
+        message: Message,
+        metadata: T,
+        prevent_eviction: bool,
```

Comment:
could this argument be a constant?

Reply:
Possibly, but there's not much point. There are database accesses, many atomic operations, and hashing operations on this path. A boolean check isn't going to matter. All this aside: the code is not exposed beyond this crate so LLVM is free to inline & optimise it if it so desires. I'd suggest that leaving it up to LLVM is the more sensible route rather than unintentionally inflating the code cache when it's not optimal to do so by monomorphising multiple instances of what amounts to the same code.